File: app/src/main/java/com/marknkamau/justjava/utils/FirebaseAuthUtils.java
package com.marknkamau.justjava.utils;


import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

public class FirebaseAuthUtils {
    private static FirebaseAuth firebaseAuth;

    static {
        firebaseAuth = FirebaseAuth.getInstance();
    }

    public static FirebaseUser getCurrentUser(){
        return firebaseAuth.getCurrentUser();
    }

    public static void logOut(){
        firebaseAuth.signOut();
    }

    public static void signIn(String email, String password, final AuthActionListener listener){
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnSuccessListener(new OnSuccessListener<AuthResult>() {
            @Override
            public void onSuccess(AuthResult authResult) {
                listener.actionSuccessful("Sign in successful");
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                listener.actionFailed(e.getMessage());
            }
        });
    }

    public static void sendPasswordResetEmail(String email, final AuthActionListener listener){
        firebaseAuth.sendPasswordResetEmail(email).addOnSuccessListener(new OnSuccessListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                listener.actionSuccessful("Password reset email sent");
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                listener.actionFailed(e.getMessage());
            }
        });
    }

    public interface AuthActionListener{
        void actionSuccessful(String response);
        void actionFailed(String response);
    }

}

File: app/src/main/java/com/marknkamau/justjava/utils/FirebaseAuthUtils.kt
package com.marknkamau.justjava.utils

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest

object FirebaseAuthUtils {
    private var firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()

    val currentUser: FirebaseUser?
        get() = firebaseAuth.currentUser

    fun logOut() {
        firebaseAuth.signOut()
    }

    fun signIn(email: String, password: String, listener: AuthActionListener) {
        firebaseAuth.signInWithEmailAndPassword(email, password)
                .addOnSuccessListener { listener.actionSuccessful("Sign in successful") }
                .addOnFailureListener { exception -> listener.actionFailed(exception.message) }
    }

    fun sendPasswordResetEmail(email: String, listener: AuthActionListener) {
        firebaseAuth.sendPasswordResetEmail(email)
                .addOnSuccessListener { listener.actionSuccessful("Password reset email sent") }.
                addOnFailureListener { exception -> listener.actionFailed(exception.message) }
    }

    fun createUser(email: String, password: String, listener: AuthActionListener) {
        firebaseAuth.createUserWithEmailAndPassword(email, password)
                .addOnSuccessListener { listener.actionSuccessful("User created successfully") }
                .addOnFailureListener { exception -> listener.actionFailed(exception.message) }
    }

    fun setUserDisplayName(name: String, listener: AuthActionListener) {
        val profileUpdate = UserProfileChangeRequest.Builder().setDisplayName(name).build()
        currentUser!!.updateProfile(profileUpdate)
                .addOnSuccessListener { listener.actionSuccessful("User display name set") }
                .addOnFailureListener { exception -> listener.actionFailed(exception.message) }
    }

    interface AuthActionListener {
        fun actionSuccessful(response: String)

        fun actionFailed(response: String?)
    }

}
